src/Aspire.Dashboard/Components/Dialogs/TextVisualizerDialog.razor.cs

```diff
@@ -0,0 +1,256 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Text;
+using System.Text.Json;
+using System.Xml;
+using System.Xml.Linq;
+using Aspire.Dashboard.Model;
+using Aspire.Dashboard.Model.Otlp;
+using Microsoft.AspNetCore.Components;
+using Microsoft.FluentUI.AspNetCore.Components;
+using Microsoft.JSInterop;
+
+namespace Aspire.Dashboard.Components.Dialogs;
+
+public partial class TextVisualizerDialog : ComponentBase, IAsyncDisposable
+{
+    // xml and json are language names supported by highlight.js
+    public const string XmlFormat = "xml";
+    public const string JsonFormat = "json";
+    public const string PlaintextFormat = "plaintext";
+
+    private readonly string _copyButtonId = $"copy-{Guid.NewGuid():N}";
+    private readonly string _openSelectFormatButtonId = $"select-format-{Guid.NewGuid():N}";
+    private readonly string _logContainerId = $"log-container-{Guid.NewGuid():N}";
+
+    private IJSObjectReference? _jsModule;
+    private List<SelectViewModel<string>> _options = null!;
+    private string? _currentValue;
+    private string _formattedText = string.Empty;
+
+    public HashSet<string?> EnabledOptions { get; } = [];
+    public string FormattedText
+    {
+        get => _formattedText;
+        private set
+        {
+            _formattedText = value;
+            FormattedLines = GetLines();
+        }
+    }
+    public ICollection<StringLogLine> FormattedLines { get; set; } = [];
+
+    public string FormatKind { get; private set; } = PlaintextFormat;
+
+    [Inject]
+    public required IJSRuntime JS { get; init; }
+
+    [Inject]
+    public required ThemeManager ThemeManager { get; init; }
+
+    protected override async Task OnInitializedAsync()
+    {
+        await ThemeManager.EnsureEffectiveThemeAsync();
+    }
+
+    protected override async Task OnAfterRenderAsync(bool firstRender)
+    {
+        if (firstRender)
+        {
+            _jsModule = await JS.InvokeAsync<IJSObjectReference>("import", "/Components/Dialogs/TextVisualizerDialog.razor.js");
+        }
+
+        if (_jsModule is not null && FormatKind != PlaintextFormat)
+        {
+            await _jsModule.InvokeVoidAsync("connectObserver", _logContainerId);
```

Comment:
This looks like it can be called multipled times if the control re-renders. Is that ok?
Should the observer still be running if the format switches back to plain text?

Reply:
> This looks like it can be called multipled times if the control re-renders. Is that ok?

It's necessary to re-highlight and re-observe if any text/the format has changed, so I'd say yes it's okay. However, the previous observer could be disconnected during subsequent calls. I made this change

> Should the observer still be running if the format switches back to plain text?

No, made that change.